tests/nested/manual/grade-signed-cloud-init-testkeys/task.yaml

```diff
@@ -0,0 +1,167 @@
+  # mkdir -p pc-gadget/meta/hooks/
+  # cp prepare-device pc-gadget/meta/hooks/
+
+  nested_secboot_sign_gadget pc-gadget "$SNAKEOIL_KEY" "$SNAKEOIL_CERT"
+  snap pack pc-gadget/ extra-snaps/
+  rm -rf pc-gadget/
+
+  rm -f "$SNAKEOIL_KEY" "$SNAKEOIL_CERT"
+
+  # TODO: enable when ready
+  # start fake device svc
+  # systemd_create_and_start_unit fakedevicesvc "$(command -v fakedevicesvc) localhost:11029"
+
+  # first boot - legit NoCloud usage
+  nested_build_seed_cdrom "$TESTSLIB/cloud-init-seeds/normal-user" seed.iso cidata user-data meta-data
+
+  # second boot - attacker drive
+  nested_build_seed_cdrom "$TESTSLIB/cloud-init-seeds/attacker-user" seed2.iso cidata user-data meta-data
+
+  # first boot will use seed1 to create the normal-user in addition to the 
+  # system-user assertion
+  NESTED_PARAM_CD="-cdrom $(pwd)/seed.iso"
+
+  nested_create_core_vm
+  nested_start_core_vm
+
+restore: |
+  if [ "$TRUST_TEST_KEYS" = "false" ]; then
+      echo "This test needs test keys to be trusted"
+      exit
+  fi
+
+  #shellcheck source=tests/lib/store.sh
+  . "$TESTSLIB"/store.sh
+  teardown_fake_store "$NESTED_FAKESTORE_BLOB_DIR"
+
+execute: |
+  if [ "$TRUST_TEST_KEYS" = "false" ]; then
+      echo "This test needs test keys to be trusted"
+      exit
+  fi
+
+  #shellcheck source=tests/lib/nested.sh
+  . "$TESTSLIB/nested.sh"
+
+  echo "The initial cloud-init user was created"
+  nested_exec "cat /var/lib/extrausers/passwd" | MATCH normal-user
+
+  echo "And we can run things as the normal user"
+  nested_exec_as normal-user ubuntu "sudo true"
+
+  # TODO: check that we got a serial assertion via the fakedevicesvc
+  # for now we just don't get a serial assertion which is fine for the purposes
+  # of this test
+
+  # TODO: is there a better thing we can wait for here instead? maybe the log
+  # message from snapd directly via journalctl ?
+  echo "Waiting for snapd to react to cloud-init"
+  sleep 60
+
+  echo "Ensuring that cloud-init got disabled after running"
+  nested_exec "cloud-init status" | MATCH "status: disabled"
+  nested_exec "test -f /etc/cloud/cloud-init.disabled"
+  nested_exec "! test -f /etc/cloud/cloud.cfg.d/zzzz_snapd.cfg"
+
+  # gracefully shutdown so that we don't have file corruption
+  echo "Gracefully shutting down the nested VM to prepare a simulated attack"
+  boot_id="$(nested_get_boot_id)"
+  nested_shutdown
+
+  # replace the seed.iso with the new attacker iso
+  mv seed2.iso seed.iso
+
+  echo "Restarting nested VM with attacker cloud-init CD-ROM drive"
+  nested_force_start_vm
+  nested_wait_for_reboot "${boot_id}"
+
+  echo "The cloud-init attacker user was not created"
+  nested_exec "cat /var/lib/extrausers/passwd" | NOMATCH attacker-user
+
+  echo "cloud-init is still disabled"
+  nested_exec "cloud-init status" | MATCH "status: disabled"
+  nested_exec "test -f /etc/cloud/cloud-init.disabled"
+  nested_exec "! test -f /etc/cloud/cloud.cfg.d/zzzz_snapd.cfg"
```

Comment:
Neat way to avoid the issue with `!` while using a function `nested_exec` which cannot be used with `not`.